Add open file dialog and example app for dialogs

obulat opened the pull request:
<!--- Describe your changes in detail -->
I added `open_file_dialog` to the core `main_window` class, as well as Windows implementation for `open_file_dialog` and `save_file_dialog`. There are several things I'm not sure about: 

1. Windows allows multiselect when opening files. Should we add it?
2. What is the format of `file_types` sent to these dialogs? Is it something like `['txt', 'py']` or `['.txt', '.py']`?
3. Save file dialog should only provide the path where to save, saving itself should be done in the app, right? And the same is true with open file dialog, it only returns the full file name.
4. If the dialog is canceled, it returns `None`. Is this preferred behavior?

<!--- What problem does this change solve? -->
<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented:
@obulat This is looking good! Regarding your questions:

1. I'd say yes, we should allow multiselect on open - but it should be an option, with "single select" being the default.
2. Filetypes should be `['txt']`, following the example of document types on toga.App.
3. Correct - the save dialog itself doesn't do the save, it's currently just a prompt for the filename. 
4. A return value of None probably isn't ideal; raising a "no file name provided" exception would allow for easier default handling, I suspect.

obulat commented:
Added multiselect as an option to `open_file_dialog`.
Tried to add filetypes to Winforms backend (needs more work, possibly in a future PR).
Added ValueError if a filename is not provided, both to the `open_file_dialog` and `save_file_dialog`.

goanpeca commented on src/core/toga/window.py:
```diff
@@ -257,3 +258,31 @@ def save_file_dialog(self, title, suggested_filename, file_types=None):
         """
         return self._impl.save_file_dialog(title, suggested_filename, file_types)
+
+    def open_file_dialog(self, title, initial_directory=None, file_types=None, multiselect=False):
```
It is usually possible on some systems to include a text associated to the extensions, for instance

`Text Files (*.txt)` or `Awesome Text files (*.txt)`

So `file_types` could be expanded to include those cases?

freakboy3742 replied:
Sounds like a good idea to me. It might be easiest to create a light "datatype" class wrapper to capture these descriptions so we can use either a list or strings or a list of "datatype" objects.
